prov/gni/include/gnix_tags.h

```diff
@@ -30,6 +30,95 @@
  * SOFTWARE.
  */
 
+/**
+ * Notes:
+ *
+ * This tag matching system currently implements a linked-list version of
+ * a tag matcher.
+ *
+ * A hash list implementation was planned but will be not pursued due to the
+ * constraints of the problem.
+ *
+ * As understood at the time of this writing, matching a tag involves matching
+ * the bits of the tag against the bits of the tag to find, less the ignored
+ * bits. When no bits are ignored, there is no subset of tags to match other
+ * than the exact match. When some bits are ignored, there is a subset of tags
+ * that are not exact matches, but are considered matches according to the
+ * ignored bits. This problem represents a search over a k-space, where k is
+ * the number of distinct ignore fields given by the tag format for the
+ * provider. Each ignore field in the tag format distinct separates tags into
+ * different 'bins' within the same field.
+ *
+ * A hash list implementation is not impossible, but can be computationally
+ * impractical. This is due to the way a search must occur if bits are ignored.
+ * The problem would be trivial if no bits were ignored, as the implementation
+ * could simply go to the bucket where the tag was hashed and search there.
+ * However, because ignored bits could hash to multiple, if not all, buckets,
+ * the tag matcher must search all buckets. It takes more time to compute all
+ * permutations of affected tags, than to just search each tag, so the default
+ * behavior is to search all buckets if a non-zero ignore field is provided.
+ *
+ * In the event that the ignore field is always zero, a hash list
+ * implementation is strictly better than a list implementation. However,
+ * once ignore bits are considered, the problem becomes more complex.
+ *
+ * Consider a hash list implementation where tags are split into K buckets,
+ * and tags are evenly distributed. If N tags are placed into the hash list in
+ * an evenly distributed manner, we can compute a best, average and worst
+ * case expectation for the hash list implementation.
+ *
+ * In the best case, the tag we are searching for is at the front of the list
+ * and only one tag in the system could match the search parameters, then
+ * we get the following best case analysis:
+ *
+ * When tags are evenly distributed, there should be ~(N/K) tags per bucket.
+ *
```

Comment:
Is there a difference between `N` and `n` (or `K` and `k`)?  There doesn't seem to be, in which case I'd be consistent and use one or the other.